chainexchange/pubsub.go

```diff
@@ -0,0 +1,306 @@
+package chainexchange
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/filecoin-project/go-f3/gpbft"
+	"github.com/filecoin-project/go-f3/internal/psutil"
+	"github.com/filecoin-project/go-f3/merkle"
+	lru "github.com/hashicorp/golang-lru/v2"
+	logging "github.com/ipfs/go-log/v2"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+var (
+	log = logging.Logger("f3/chainexchange")
+
+	_ ChainExchange      = (*PubSubChainExchange)(nil)
+	_ pubsub.ValidatorEx = (*PubSubChainExchange)(nil).validatePubSubMessage
+
+	chainPortionPlaceHolder = &chainPortion{}
+)
+
+type chainPortion struct {
+	// chain is a pointer to the full chain in order to avoid copying the chain across the cache
+	fullChain *gpbft.ECChain
+	offset    int
```

Comment:
ECChain is a slice, so there is no reason to make it a pointer (I guess it saves a few bytes, but I don't think a few bytes will break us), as the slice is a tuple of `[array_pointer, capacity, length]`
You could also re-slice the fullChain instead of saving offset.